Lib/unittest/mock.py

```diff
@@ -737,14 +737,47 @@ def _format_mock_failure_message(self, args, kwargs):
         return message % (expected_string, actual_string)
 
 
+    def _get_call_signature_from_name(self, name):
+        """
+        If there is no name like the case for call check against functions
+        with tuples then return the self's spec signature.
+        If the name is not empty then remove () and split by '.' to get
+        list of names to iterate through the children until a potential
+        match is found.
+        """
+        if not name:
+            return self._spec_signature
+
+        sig = None
+        names = name.replace('()', '').split('.')
+        children = self._mock_children
+
+        for name in names:
+            child = children.get(name)
+            if child is None:
+                break
+            else:
+                children = child._mock_children.get(name)
+                sig = child._spec_signature
+                if children is None:
+                    break
+
+        return sig
+
+
     def _call_matcher(self, _call):
         """
         Given a call (or simply an (args, kwargs) tuple), return a
         comparison key suitable for matching with other calls.
         This is a best effort method which relies on the spec's signature,
         if available, or falls back on the arguments themselves.
         """
-        sig = self._spec_signature
+
+        if isinstance(_call, tuple) and len(_call) > 2:
```

Comment:
Yes, any sequence can be used. It should be a two element sequence args and kwargs or three element sequence in case of name being present. But most examples and tests construct `call` which inherits from tuple or tuple as a recommended one.

```
>>> from unittest.mock import Mock
>>> m = Mock()
>>> m(1)
<Mock name='mock()' id='4554182024'>
>>> m.assert_has_calls([[(1,), {}]])
```

Reply:
This feels like it would be sensible to just state and ensure that these are only ever `_Call` objects.

Reply:
@tirkarthi - thoughts on my comment here ^^

Reply:
Though this is an internal method there is a note on the docstring of `_call_matcher` that it could be a tuple. It's not specified in docs that this could be a tuple. I haven't seen a usecase of it in other projects too. There is a test for this usecase but it's not using autospec so it didn't fail with my PR I guess. Relevant test using tuples https://github.com/python/cpython/blob/9fe42b49c79c453d905d0395150ba0607fbab18b/Lib/unittest/test/testmock/testmock.py#L1265

```
Given a call (or simply an (args, kwargs) tuple), return a
comparison key suitable for matching with other calls.
```

There could be a below test that passes with checking `_call` to be a tuple.

```python
    def test_assert_has_calls_tuple(self):
        class Something:

            def __init__(self): pass
            def meth(self, a, b, c, d=None): pass

        m = create_autospec(Something)
        m.meth(1, 2, 3, 1)
        m.assert_has_calls([('meth', (1, 2, 3, 1), {})])
```

On having a strict check of only using `_Call` it will fail since it will go on to use the constructor signature.

```
./python.exe -m unittest Lib.unittest.test.testmock.testmock.MockTest.test_assert_has_calls_tuple
F
======================================================================
FAIL: test_assert_has_calls_tuple (Lib.unittest.test.testmock.testmock.MockTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/unittest/test/testmock/testmock.py", line 1387, in test_assert_has_calls_tuple
    m.assert_has_calls([('meth', (1, 2, 3, 1), {})])
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/unittest/mock.py", line 876, in assert_has_calls
    raise AssertionError(
AssertionError: Calls not found.
Expected: [('meth', (1, 2, 3, 1), {})]
Actual: [call.meth(1, 2, 3, 1)].

----------------------------------------------------------------------
Ran 1 test in 0.063s

FAILED (failures=1)
```

I am leaning towards removing tuple to support only `_Call` since the error message is also not useful. We can perhaps change it in 3.9 to accept only `_Call` and then revert back if someone complains that it breaks their workflow given there is plenty of time. I am not sure about backporting it to 3.8 though since beta is out. Would also like to know thoughts from @voidspace since this was added in the original commit where mock was added.

Reply:
Ah, if it being `_Call`-only causes problems then that's probably a good reason to leave it as it is..

Reply:
Splitting hairs: should the ` and len(_call) > 2` move inside `self._get_call_signature_from_name` and maybe renamed that method `_get_signature_from_call`?